keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,641 @@
+warning:
+
+```text
+v1 Endpoints is deprecated in v1.33+; use discovery.k8s.io/v1 EndpointSlice
+```
+
+### Warnings via metrics
+
+We will add a metric to the apiserver counting the number of Endpoints
+API operations, labeled by service account name, to help
+administrators find and update clients that are still using Endpoints.
+
+To avoid cardinality problems, we will only add a fixed number of
+labels, and just list further clients as something like "other" or
+"additional clients".
+
+### Documentation Updates
+
+A few examples in the official documentation still need to be updated
+to use EndpointSlices rather than Endpoints.
+
+### `kubernetes.default` Endpoints
+
+Both the Endpoints and the EndpointSlice for the `kubernetes.default`
+Service are created by the apiserver rather than by
+kube-controller-manager (and are thus independent of whether the
+Endpoints controller is running or not).
+
+For now, we will not change this (beyond that anyone reading the
+object will now see the deprecation warning).
+
+If, in the future, we decide to disable the Endpoints controller by
+default, we can consider whether it makes sense to stop creating the
+apiserver Endpoints as well. It is possible that we could decide that
+there is a stronger "API" guarantee around the existence of the
+`kubernetes` Endpoints object than there is around Endpoints objects
+in general.
+
+### Endpoints Cleanup
+
+We do not want to leave stale `Endpoints` objects around forever if
+the Endpoints controllers are not running. (This is both a waste of
+disk space and a potential source of confusion since the Endpoints
+objects would quickly become out-of-date and incorrect.)
+
+One possibility would be to just document that administrators should
+delete all existing Endpoints themselves if they are going to disable
+the controllers.
+
+Another possibility would be to create an `endpoints-cleanup`
+controller, that could be enabled explicitly, and document that admins
+should (probably) enable that controller if they disable the others.
+
+Or perhaps the EndpointSlice controller could delete Endpoints objects
+that were more than 24 hours out of date with respect to their
+EndpointSlices?
+
+In all cases, we should probably not automatically delete Endpoints
+that don't looke like they were originally created by the Endpoints
+controller. (That is, we should not delete Endpoints unless they
+correspond to a Service with a selector.)
+
+```
+<<[UNRESOLVED endpoints-cleanup ]>>
+
+Decide what to do here. (In the earlier stages we can just recommend
+manual deletion.)
+
+<<[/UNRESOLVED]>>
+```
+
+### Update Remaining Internal Endpoints Users
+
+The aggregated API server and apiserver service proxying code still
+make use of Endpoints. They will need to be updated to use
+EndpointSlices (and, perhaps, to fall back to Endpoints for a while,
+in case any users were manually creating Endpoints and not mirroring
+them?).
```

Comment:
Would endpoint-slice-mirror controller mirror them in such case?

Reply:
Probably, yes. FYI, there's already a PR for this (https://github.com/kubernetes/kubernetes/pull/129837) and the release note is currently:

```
When proxying to an aggregated API server, kube-apiserver now uses the
EndpointSlices of the `service` indicated by the `APIServer`, rather than
using Endpoints.

If you are using the aggregated API server feature, and you are writing out
the endpoints for it by hand (rather than letting kube-controller-manager
generate Endpoints and EndpointSlices for it automatically based on the
Service definition), then you should write out an EndpointSlice object rather
than (or in addition to) an Endpoints object.
```